# 1
Utilize o conjunto de dados [Go To College Dataset](https://www.kaggle.com/datasets/saddamazyazy/go-to-college-dataset) ou um conjunto de dados de **classificação supervisionada** de sua escolha.

In [29]:
import pickle
import pandas as pd

In [30]:
with open("clf.pkl", "rb") as files:
    clf = pickle.load(files)

clf

KNeighborsClassifier(n_neighbors=10)

In [31]:
with open('X.pkl', 'rb') as files:
    X =pickle.load(files)

X 

type_school_encoded  school_accreditation_encoded  gender_encoded  \
0                      0                             0               1   
1                      0                             0               1   
2                      0                             1               0   
3                      1                             1               1   
4                      0                             0               0   
..                   ...                           ...             ...   
995                    1                             0               0   
996                    0                             1               0   
997                    1                             0               1   
998                    0                             1               1   
999                    0                             1               1   

     interest_encoded  residence_encoded  parent_was_in_college_encoded  
0                   1                  1                              0  
1                   1                  1                              0  
2                   4                  1                              0  
3                   4                  0                              1  
4                   4                  1                              0  
..                ...                ...                            ...  
995                 4                  0                              1  
996                 1                  0                              1  
997                 1                  1                              0  
998                 3                  0                              1  
999                 0                  1                              1  

[1000 rows x 6 columns]

In [32]:
with open('y.pkl', 'rb') as files:
    y = pickle.load(files)

y 

[1,
 1,
 1,
 1,
 0,
 0,
 1,
 0,
 1,
 0,
 0,
 0,
 0,
 1,
 0,
 1,
 0,
 0,
 0,
 0,
 1,
 0,
 0,
 0,
 1,
 1,
 0,
 1,
 0,
 0,
 1,
 0,
 0,
 0,
 1,
 1,
 1,
 0,
 0,
 0,
 1,
 0,
 1,
 1,
 1,
 0,
 0,
 1,
 1,
 0,
 1,
 1,
 1,
 0,
 0,
 1,
 1,
 0,
 0,
 1,
 1,
 0,
 1,
 0,
 0,
 1,
 0,
 0,
 1,
 1,
 1,
 0,
 0,
 0,
 1,
 0,
 1,
 1,
 1,
 1,
 0,
 0,
 1,
 1,
 0,
 0,
 1,
 1,
 1,
 1,
 0,
 0,
 1,
 0,
 1,
 1,
 0,
 1,
 1,
 1,
 0,
 0,
 1,
 0,
 1,
 0,
 1,
 1,
 0,
 0,
 0,
 0,
 1,
 0,
 1,
 0,
 0,
 0,
 1,
 0,
 0,
 1,
 1,
 1,
 1,
 0,
 1,
 0,
 1,
 0,
 0,
 0,
 1,
 0,
 1,
 0,
 1,
 0,
 0,
 0,
 1,
 0,
 0,
 1,
 0,
 1,
 1,
 0,
 1,
 0,
 1,
 1,
 1,
 1,
 0,
 1,
 0,
 0,
 0,
 0,
 1,
 1,
 0,
 0,
 1,
 1,
 0,
 0,
 1,
 0,
 1,
 1,
 0,
 1,
 0,
 0,
 1,
 1,
 1,
 0,
 0,
 1,
 0,
 1,
 0,
 1,
 1,
 0,
 1,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 0,
 1,
 0,
 1,
 0,
 1,
 0,
 0,
 1,
 0,
 0,
 1,
 0,
 1,
 0,
 0,
 1,
 0,
 1,
 1,
 1,
 0,
 1,
 1,
 1,
 0,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 0,
 1,
 0,
 0,
 1,
 1,
 1,
 1,
 0,
 0,
 0,
 1,
 1,
 1,
 1,
 0,
 1,


# 2
Utilize a técnica k-fold cross validation. Compare os resultados com o melhor resultado da Atividade 6.

In [33]:
from sklearn.model_selection import cross_validate
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import recall_score,precision_score, f1_score, accuracy_score, classification_report

In [34]:
scores = cross_validate(clf, X, y, scoring=['accuracy','f1','recall'],cv=10)
scores

{'fit_time': array([0.00409079, 0.00267744, 0.00300241, 0.00223541, 0.00399399,
        0.00707412, 0.00291085, 0.00308728, 0.00292563, 0.00291944]),
 'score_time': array([0.00674176, 0.0055542 , 0.00757051, 0.00806451, 0.00800872,
        0.00701618, 0.00605941, 0.00500274, 0.00608301, 0.00499821]),
 'test_accuracy': array([0.63, 0.68, 0.66, 0.61, 0.75, 0.69, 0.65, 0.7 , 0.67, 0.71]),
 'test_f1': array([0.59340659, 0.63636364, 0.63043478, 0.54117647, 0.75728155,
        0.64367816, 0.62365591, 0.65909091, 0.64516129, 0.69473684]),
 'test_recall': array([0.54, 0.56, 0.58, 0.46, 0.78, 0.56, 0.58, 0.58, 0.6 , 0.66])}

In [35]:
print(f"Resultado usando k-fold: {scores['test_accuracy'].mean()} +- {scores['test_accuracy'].std()}")
print(f"Tempo de treinamento usando k-fold: {scores['fit_time'].mean()} +- {scores['fit_time'].std()}")
print(f"Tempo de teste usando k-fold: {scores['score_time'].mean()} +- {scores['score_time'].std()}")

Resultado usando k-fold: 0.675 +- 0.038535697735995376
Tempo de treinamento usando k-fold: 0.0034917354583740234 +- 0.0013066925401404774
Tempo de teste usando k-fold: 0.006509923934936523 +- 0.0010933217803666127


No modelo Hold out no exercicio 6 a acurácia foi 95%, já no K-fold foi de 98%, o que torna o K-fold mais preciso.





# 3
Realize experimentos utilizando normalização e padronização. Compare os resultados com o melhor resultado da Atividade 6.

In [36]:
X

type_school_encoded  school_accreditation_encoded  gender_encoded  \
0                      0                             0               1   
1                      0                             0               1   
2                      0                             1               0   
3                      1                             1               1   
4                      0                             0               0   
..                   ...                           ...             ...   
995                    1                             0               0   
996                    0                             1               0   
997                    1                             0               1   
998                    0                             1               1   
999                    0                             1               1   

     interest_encoded  residence_encoded  parent_was_in_college_encoded  
0                   1                  1                              0  
1                   1                  1                              0  
2                   4                  1                              0  
3                   4                  0                              1  
4                   4                  1                              0  
..                ...                ...                            ...  
995                 4                  0                              1  
996                 1                  0                              1  
997                 1                  1                              0  
998                 3                  0                              1  
999                 0                  1                              1  

[1000 rows x 6 columns]

In [37]:
df = X[['type_school_encoded', 'school_accreditation_encoded', 'gender_encoded', 'interest_encoded',
       'residence_encoded', 'parent_was_in_college_encoded']]

df

type_school_encoded  school_accreditation_encoded  gender_encoded  \
0                      0                             0               1   
1                      0                             0               1   
2                      0                             1               0   
3                      1                             1               1   
4                      0                             0               0   
..                   ...                           ...             ...   
995                    1                             0               0   
996                    0                             1               0   
997                    1                             0               1   
998                    0                             1               1   
999                    0                             1               1   

     interest_encoded  residence_encoded  parent_was_in_college_encoded  
0                   1                  1                              0  
1                   1                  1                              0  
2                   4                  1                              0  
3                   4                  0                              1  
4                   4                  1                              0  
..                ...                ...                            ...  
995                 4                  0                              1  
996                 1                  0                              1  
997                 1                  1                              0  
998                 3                  0                              1  
999                 0                  1                              1  

[1000 rows x 6 columns]

In [38]:
X_train,X_test,y_train,y_test = train_test_split(X, y, train_size=0.9, random_state=1)
clf = KNeighborsClassifier(n_neighbors=7)
clf.fit(X_train,y_train)
y_pred = clf.predict(X_test)
accuracy_score(y_test,y_pred)

0.54

In [39]:
from sklearn.preprocessing import StandardScaler
scaleX = StandardScaler()
X_train = scaleX.fit_transform(X_train)
X_test = scaleX.fit_transform(X_test)

X_train

array([[-0.79772404, -1.05014463,  0.97151659,  0.37284316, -1.09566829,
         0.96290958],
       [-0.79772404,  0.95224979,  0.97151659,  1.08984922, -1.09566829,
         0.96290958],
       [-0.79772404,  0.95224979, -1.0293185 , -1.06116898,  0.91268499,
         0.96290958],
       ...,
       [ 1.25356634, -1.05014463, -1.0293185 ,  1.08984922,  0.91268499,
        -1.03851911],
       [ 1.25356634,  0.95224979,  0.97151659,  0.37284316,  0.91268499,
        -1.03851911],
       [ 1.25356634, -1.05014463, -1.0293185 ,  1.08984922, -1.09566829,
         0.96290958]])

# 4
Tente realizar alguma outra etapa pré-processamento/transformação dos dados. Compare os resultados com os resultados já obtidos.

In [40]:
from sklearn.pipeline import Pipeline

In [41]:
X_train,X_test,y_train,y_test = train_test_split(X, y, train_size=0.9, random_state=1)

In [42]:
X_train

type_school_encoded  school_accreditation_encoded  gender_encoded  \
335                    0                             0               1   
13                     0                             1               1   
911                    0                             1               0   
156                    1                             0               1   
428                    0                             0               1   
..                   ...                           ...             ...   
767                    1                             0               0   
72                     0                             1               0   
908                    1                             0               0   
235                    1                             1               1   
37                     1                             0               0   

     interest_encoded  residence_encoded  parent_was_in_college_encoded  
335                 3                  0                              1  
13                  4                  0                              1  
911                 1                  1                              1  
156                 3                  1                              0  
428                 3                  0                              1  
..                ...                ...                            ...  
767                 3                  1                              0  
72                  4                  0                              1  
908                 4                  1                              0  
235                 3                  1                              0  
37                  4                  0                              1  

[900 rows x 6 columns]

In [43]:
pipeline = Pipeline([
    ('scaler', StandardScaler()),
    ('clf', KNeighborsClassifier(n_neighbors = 7))
    ])

In [44]:
X_train

type_school_encoded  school_accreditation_encoded  gender_encoded  \
335                    0                             0               1   
13                     0                             1               1   
911                    0                             1               0   
156                    1                             0               1   
428                    0                             0               1   
..                   ...                           ...             ...   
767                    1                             0               0   
72                     0                             1               0   
908                    1                             0               0   
235                    1                             1               1   
37                     1                             0               0   

     interest_encoded  residence_encoded  parent_was_in_college_encoded  
335                 3                  0                              1  
13                  4                  0                              1  
911                 1                  1                              1  
156                 3                  1                              0  
428                 3                  0                              1  
..                ...                ...                            ...  
767                 3                  1                              0  
72                  4                  0                              1  
908                 4                  1                              0  
235                 3                  1                              0  
37                  4                  0                              1  

[900 rows x 6 columns]

In [45]:
pipeline.fit(X_train, y_train)

Pipeline(steps=[('scaler', StandardScaler()),
                ('clf', KNeighborsClassifier(n_neighbors=7))])

In [46]:
pipeline.score(X_test, y_test)

0.55

In [47]:
y_pred = pipeline.predict(X_test)
accuracy_score(y_test,y_pred)

0.55

Usando o pipeline a acurácia deu um valor maior a acuracia foi de 97% para 99%